### PR TITLE
Initial NLP implementation into dp-search-api search 

### PR DESCRIPTION
**Pull request by KamenDimitrov97**
### What

Describe what you have changed and why.

### How to review

Describe the steps required to test the changes.

### Who can review

Describe who worked on the changes, so that other people can review.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -82,19 +86,32 @@ func NewSearchAPI(router *mux.Router, dpESClient DpElasticSearcher, deprecatedES
...
+			settingsNLP,
+			cli,
 			a.dpESClient,
 			transformer,
 		),
 	).Methods(http.MethodGet)
 	return a
 }
 
+// RegisterGetSearch registers the handler for GET /nlp/search endpoint
```
Comment (philtweir) [inline api/api.go:104]: RegisterGetNlpSearch? but also the is the endpoint intended to group `nlp` related resources under `/nlp/` or is this a single endpoint, in which case, something like `search-nlp`  might be more appropriate

File: api/search.go [hunk 2]
```diff
@@ -208,13 +219,90 @@ func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryPar
...
+func SearchHandlerFunc(validator QueryParamValidator, queryBuilder QueryBuilder, nlpConfig config.NLP, nlpCLI *nlp.Client, elasticSearchClient DpElasticSearcher, transformer ResponseTransformer) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
 		ctx := req.Context()
 		params := req.URL.Query()
 
-		q, searchReq, countReq := CreateRequests(w, req, validator)
+		q := params.Get("q")
+
+		var nlpCriteria *query.NlpCriteria
+		if params.Get("c") == "1" {
```
Comment (philtweir) [inline api/search.go:292]: Is this idiomatic within dp-search-api or is there any other preferred way for testing a boolean parameter is on? (e.g. is there a comparable example elsewhere?)

File: api/search.go [hunk 3]
```diff
@@ -299,12 +387,29 @@ func SearchHandlerFunc(validator QueryParamValidator, queryBuilder QueryBuilder,
...
 
-		q, searchReq, countReq := CreateRequests(w, req, validator)
+		var nlpCriteria *query.NlpCriteria
+		q := params.Get("q")
+		if params.Get("c") == "1" {
+			nlpSettings := query.NlpSettings{}
+
+			// Load default settings
+			// FIXME: move this somewhere better
+			json.Unmarshal([]byte(nlpConfig.NlpHubSettings), &nlpSettings)
```
Comment (philtweir) [inline api/search.go:402]: Now that we have removed NLP hub - is there something that should be rearrange with NlpHubSettings or is it now just used within dp-search-api?

File: api/search.go [hunk 4]
```diff
@@ -299,12 +387,29 @@ func SearchHandlerFunc(validator QueryParamValidator, queryBuilder QueryBuilder,
...
+
+			// Load settings for this request
+			nlpSettingsRequest := params.Get("nlpSettings")
+			if nlpSettingsRequest != "" {
+				json.Unmarshal([]byte(nlpSettingsRequest), &nlpSettings)
```
Comment (philtweir) [inline api/search.go:407]: Are we OK taking the settings from the query or are there settings we should not? (are those all envs?)

File: api/search.go [hunk 5]
```diff
@@ -474,3 +579,70 @@ func processCountQuery(ctx context.Context, elasticSearchClient DpElasticSearche
...
+			log.Error(ctx, "error making request to category: %w", err)
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+
+		var err error
+
+		berlin, err = cli.GetBerlin(ctx, scrubber.Query)
```
Comment (philtweir) [inline api/search.go:612]: We'll need to have the berlin-cleaned query at some point, but fine to start

File: api/search.go [hunk 6]
```diff
@@ -474,3 +579,70 @@ func processCountQuery(ctx context.Context, elasticSearchClient DpElasticSearche
...
+	nlpResponse := models.NLPResp{
+		Berlin:   berlin,
+		Scrubber: scrubber,
+		Category: category,
+	}
+
+	var nlpCriteria *query.NlpCriteria
+	if len(nlpResponse.Category) > 0 {
+		for i, cat := range nlpResponse.Category {
+			if nlpSettings.CategoryLimit > 0 && nlpSettings.CategoryLimit <= i {
```
Comment (philtweir) [inline api/search.go:629]: Maybe a comment to explain this conditional?

File: api/search.go [hunk 7]
```diff
@@ -474,3 +579,70 @@ func processCountQuery(ctx context.Context, elasticSearchClient DpElasticSearche
...
+				break
+			}
+			log.Warn(ctx, cat.Code[0])
```
Comment (philtweir) [inline api/search.go:632]: Do we want this logging output (as Warn, anyway)?

File: features/steps/component.go [hunk 8]
```diff
@@ -88,6 +92,63 @@ func SearchAPIComponent(authFeature *componentTest.AuthorizationFeature) (c *Com
...
+	c.FakeNLPSearchAPI.fakeHTTP.NewHandler().
+		Get("/v1/berlin/search").
+		Reply(200).
+		BodyStruct(berlin)
+	c.FakeNLPSearchAPI.fakeHTTP.NewHandler().
+		Get("/v1/scrubber").
+		Reply(200).
+		BodyStruct(scrubber)
+	c.FakeNLPSearchAPI.fakeHTTP.NewHandler().
+		Get("/categories").
```
Comment (philtweir) [inline features/steps/component.go:147]: Not for this PR but we should get a `v1` in the category API route.

File: query/search.go [hunk 9]
```diff
@@ -27,21 +27,23 @@ var es710AggregationField = &AggregationFields{
...
-	AggregationField  string // Deprecated (used only in legacy templates for aggregations)
-	AggregationFields *AggregationFields
-	Highlight         bool
-	URIPrefix         string
-	Topic             []string
-	TopicWildcard     []string
-	PopulationTypes   []*PopulationTypeRequest
-	Dimensions        []*DimensionRequest
-	Now               string
+	Term                string
```
Comment (philtweir) [inline query/search.go:30]: Intended whitespace change? (aside from lines with Nlp obvs)

File: swagger.yaml [hunk 10]
```diff
@@ -592,12 +815,3 @@ definitions:
               type: array
               items:
                 type: string
-
```
Comment (philtweir) [inline swagger.yaml:595]: Should this be removed?

File: api/search.go [hunk 11]
```diff
@@ -100,7 +103,7 @@ func paramGetBool(params url.Values, key string, defaultValue bool) bool {
 
 // CreateRequests reads the parameters from the request and generates the corresponding SearchRequest and CountRequest
 // If any validation fails, the http.Error is already handled, and nil is returned: in this case the caller may return straight away
-func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryParamValidator) (string, *query.SearchRequest, *query.CountRequest) {
+func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryParamValidator, nlpCritiria *query.NlpCriteria) (string, *query.SearchRequest, *query.CountRequest) {
```
Comment (philtweir) [inline api/search.go:106]: Typo in nlpCriteria (but also shouldn't test stop this?)

Comment (philtweir) [reply api/search.go:106]: Oh, it's consistent in function

File: api/search.go [hunk 12]
```diff
@@ -167,6 +170,14 @@ func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryPar
 		Now:       time.Now().UTC().Format(time.RFC3339),
 	}
 
+	if nlpCritiria != nil && nlpCritiria.UseCategory {
+		reqSearch.NlpCategories = nlpCritiria.Categories
+	}
+
+	if nlpCritiria != nil && nlpCritiria.UseSubdivision {
+		reqSearch.NlpSubdivisionWords = nlpCritiria.SubdivisionWords
+	}
```
Comment (philtweir) [inline api/search.go:179]: We're adding `Nlp*` to the `reqSearch` struct, but is it actually being subsequently used? In other words, does it get put into part of the ES templated query to affect the returned data. At the least, running locally, it should be possible to see the impact of including a category term on the results when hitting the API with `curl` (and it's connected to a populated test ES instance).

As an example, tracking through the approach for [PopulationTypes](https://github.com/ONSdigital/dp-search-api/blob/d00812870a762b255f413a09b402a63526ba3fb6/query/templates/search/v710/populationTypeFilters.tmpl#L4) (and the ES NLP changes in the templates folder from the alpha), you should see the follow-on step -- or maybe I'm missing something else in this PR?

File: api/api.go (hunk 1 above)
Comment (KamenDimitrov97) [reply api/api.go:104]: Well the idea was to group the endpoints, but considering that there's only one, maybe I should've renamed it..
Will do now

File: query/query.go [hunk 13]
```diff
@@ -14,10 +14,30 @@ import (
...
+	Categories       []NlpCriteriaCategory
+	UseSubdivision   bool
+	SubdivisionWords string
+}
+
+type NlpSettings struct {
+	CategoryWeighting float32 `json:"categoryWeighting"`
+	CategoryLimit     int     `json:"categoryLimit"`
+	DefaultState      string  `json:"defaultState"`
+}
```
Comment (lindenmckenzie) [inline query/query.go:39]: Should these still live here?

File: query/templates/search/contentFilterOnURIPrefix.tmpl [hunk 14]
```diff
@@ -1,7 +1,7 @@
 
       {{if .URIPrefix}}
           { "prefix" :
-                { "uri" : "{{.URIPrefix}}" }
+                { "uri" : { "value": "{{.URIPrefix}}", "boost": 10.0 } }
           }
         {{if or .Topic .TopicWildcard}},{{end}}
```
Comment (lindenmckenzie) [inline query/templates/search/contentFilterOnURIPrefix.tmpl:6]: Why this change?

File: query/templates/search/contentQuery.tmpl [hunk 15]
```diff
@@ -1,11 +1,16 @@
 {
+explain: true,
```
Comment (lindenmckenzie) [inline query/templates/search/contentQuery.tmpl:2]: shouldn't 'explain' have quote marks?

Also indentation.

File: query/templates/search/contentQuery.tmpl [hunk 16]
```diff
@@ -1,11 +1,16 @@
...
  {{- if .From -}}
  	"from" : {{- .From}},
  {{- end}}
  "size" : {{.Size}},
  "query" : {
      "bool" : {
+     "should": [
+             {{template "nlpCategory.tmpl" .}}
+        ],
+
```
Comment (lindenmckenzie) [inline query/templates/search/contentQuery.tmpl:13]: if the nlp is turned off and nothing is present here, will this present a problem?

File: query/templates/search/search.tmpl [hunk 17]
```diff
@@ -1,5 +1,3 @@
 {{/* $$ indicates newline, each header and query MUST be on its own line - with a blank line at the end */}}
 {{- template "contentHeader.tmpl" .}}$$
 {{- template "contentQuery.tmpl" .}}$$
-{{- template "countHeader.tmpl" .}}$$
-{{- template "countQuery.tmpl" .}}$$
```
Comment (lindenmckenzie) [inline query/templates/search/search.tmpl:5]: Why this change?

File: query/templates/search/v710/contentQuery.tmpl [hunk 18]
```diff
@@ -6,15 +6,18 @@
...
+			],
+			"must" : {
+				{{- if .Term}}
+					{{- template "weightedQuery.tmpl" .}}
+				{{- else}}
+					{{- template "matchAll.tmpl" .}}
+				{{- end}}
+			}
+			{{template "contentFilters.tmpl" .}}
+		}
```
Comment (lindenmckenzie) [inline query/templates/search/v710/contentQuery.tmpl:20]: Identation.

File: swagger.yaml [hunk 19]
```diff
@@ -59,6 +59,28 @@ paths:
...
+      produces:
+        - application/json
+      responses:
+        200:
+          description: OK
+          schema:
+            $ref: "#/definitions/Hub"
+        500:
+          $ref: '#/responses/InternalError'
+
```
Comment (lindenmckenzie) [inline swagger.yaml:83]: what's this for?

**Review by lindenmckenzie — COMMENTED** [review]
Some comments - sorry I didn't submit before...
